internal/upstreamldap/upstreamldap.go

```diff
@@ -0,0 +1,383 @@
+	CABundle []byte
+
+	// BindUsername is the username to use when performing a bind with the upstream LDAP IDP.
+	BindUsername string
+
+	// BindPassword is the password to use when performing a bind with the upstream LDAP IDP.
+	BindPassword string
+
+	// UserSearch contains information about how to search for users in the upstream LDAP IDP.
+	UserSearch UserSearchConfig
+
+	// Dialer exists to enable testing. When nil, will use a default appropriate for production use.
+	Dialer LDAPDialer
+}
+
+// UserSearchConfig contains information about how to search for users in the upstream LDAP IDP.
+type UserSearchConfig struct {
+	// Base is the base DN to use for the user search in the upstream LDAP IDP.
+	Base string
+
+	// Filter is the filter to use for the user search in the upstream LDAP IDP.
+	Filter string
+
+	// UsernameAttribute is the attribute in the LDAP entry from which the username should be
+	// retrieved.
+	UsernameAttribute string
+
+	// UIDAttribute is the attribute in the LDAP entry from which the user's unique ID should be
+	// retrieved.
+	UIDAttribute string
+}
+
+type Provider struct {
+	c ProviderConfig
+}
+
+// Create a Provider. The config is not a pointer to ensure that a copy of the config is created,
+// making the resulting Provider use an effectively read-only configuration.
+func New(config ProviderConfig) *Provider {
+	return &Provider{c: config}
+}
+
+// A reader for the config. Returns a copy of the config to keep the underlying config read-only.
+func (p *Provider) GetConfig() ProviderConfig {
+	return p.c
+}
+
+func (p *Provider) dial(ctx context.Context) (Conn, error) {
+	hostAndPort, err := hostAndPortWithDefaultPort(p.c.Host, ldap.DefaultLdapsPort)
+	if err != nil {
+		return nil, ldap.NewError(ldap.ErrorNetwork, err)
+	}
+	if p.c.Dialer != nil {
+		return p.c.Dialer.Dial(ctx, hostAndPort)
+	}
+	return p.dialTLS(ctx, hostAndPort)
+}
+
+// dialTLS is the default implementation of the Dialer, used when Dialer is nil.
+// Unfortunately, the go-ldap library does not seem to support dialing with a context.Context,
+// so we implement it ourselves, heavily inspired by ldap.DialURL.
+func (p *Provider) dialTLS(ctx context.Context, hostAndPort string) (Conn, error) {
+	var rootCAs *x509.CertPool
+	if p.c.CABundle != nil {
+		rootCAs = x509.NewCertPool()
+		if !rootCAs.AppendCertsFromPEM(p.c.CABundle) {
+			return nil, ldap.NewError(ldap.ErrorNetwork, fmt.Errorf("could not parse CA bundle"))
+		}
+	}
+
+	dialer := &tls.Dialer{Config: &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		RootCAs:    rootCAs,
+	}}
+
+	c, err := dialer.DialContext(ctx, "tcp", hostAndPort)
+	if err != nil {
+		return nil, ldap.NewError(ldap.ErrorNetwork, err)
+	}
+
+	conn := ldap.NewConn(c, true)
+	conn.Start()
+	return conn, nil
+}
+
+// Adds the default port if hostAndPort did not already include a port.
+func hostAndPortWithDefaultPort(hostAndPort string, defaultPort string) (string, error) {
```

Comment:
This func would be much nicer if host and port were just separate fields in the API.

Reply:
I could go either way.

Reply:
Moved to #612